orm_lib/inc/drogon/orm/DbClient.h

```diff
@@ -14,7 +14,7 @@
 
 #pragma once
 
-#if defined(USE_POSTGRESQL) && !USE_POSTGRESQL && !USE_MYSQL && !USE_SQLITE3
+#if !USE_ORM
```

Comment:
Please remove the Options for building drogon library in public headers for building applications.

Reply:
Because the USE_ORM macro is not defined when users build applications.

Reply:
I did this because when not building orm, the headers defines some method that does not have any implementation (as the .cpp files are not compiled). 

I could add a define in config.h: `#cmakedefine01 USE_ORM` (same for Redis), is it an acceptable solution ?

Reply:
For ORM classes, all methods have been implemented whether the USE_ORM is true or false, and for Redis, I'll add a 'RedisClientSkipped.cpp' file for it if Hiredis is not found in the system. (You could find the DbClientManagerSkipped.cc and RedisClientManagerSkipped.cc in Drogon for the same purpose)
BTW, the config.h is not installed in the system, it is only available when building Drogon.

Reply:
Ok, i'll take deeper look at this tomorrow then

Reply:
@Thanks so much, about the Macros, here is an explanation:
https://github.com/an-tao/trantor/pull/40#discussion_r307767558

Reply:
There might be something I don't understand :pensive: 

Let's say `BUILD_ORM` is set to `OFF`

The `HttpAppFramework.h` includes `drogon/orm/DbClient.h` from where all symbols of ORM are defined by following all `.h` files in orm folder (`DbClient`, `SQLError`, `ArrayParser`, `Result`...).

But, in the `CMakeLists.txt` all the corresponding .cc files are exluded: 

```
if (pg_FOUND OR MySQL_FOUND OR SQLite3_FOUND)
    set(DROGON_SOURCES
            ${DROGON_SOURCES}
            orm_lib/src/ArrayParser.cc
            orm_lib/src/Criteria.cc
            orm_lib/src/DbClient.cc
            orm_lib/src/DbClientImpl.cc
            orm_lib/src/DbClientLockFree.cc
            orm_lib/src/DbClientManager.cc
            orm_lib/src/DbConnection.cc
            orm_lib/src/Exception.cc
            orm_lib/src/Field.cc
            orm_lib/src/Result.cc
            orm_lib/src/Row.cc
            orm_lib/src/SqlBinder.cc
            orm_lib/src/TransactionImpl.cc
            orm_lib/src/RestfulController.cc)
else (pg_FOUND OR MySQL_FOUND OR SQLite3_FOUND)
    set(DROGON_SOURCES ${DROGON_SOURCES} lib/src/DbClientManagerSkipped.cc)
endif (pg_FOUND OR MySQL_FOUND OR SQLite3_FOUND)
```

I see that in `DbClient.cc` the method is implemented using the `USE_POSTGRESQL` macro, but as the `DbClient.cc` file is not included, the implementation is not compiled. Thus the linker can complain. It seems that gcc does not raise an error, but msvc does... And with my understanding, I think here msvc is right...

Am I missing something ?

Reply:
Yes, to avoid errors in msvc, We could add a DbClientSkipped.cc as the one of DbClientManager.
I can help you add it.

Reply:
That is not only DbClient, but also ArrayParser, Exception, Field, Result, Row... ?

Reply:
I removed the macro from public headers as requested, and I will write some unit tests to have the linker error with gcc as I have with msvc. As these symbols are unused, gcc does not complain until you use them somewhere.

Reply:
If we enable `db_test` even with `BUILD_ORM` to `OFF`, we get the linker error

Reply:
> That is not only DbClient, but also ArrayParser, Exception, Field, Result, Row... ?

How about this:
```cmake
set(DROGON_SOURCES
            ${DROGON_SOURCES}
            orm_lib/src/ArrayParser.cc
            orm_lib/src/Criteria.cc
            orm_lib/src/DbClient.cc
            orm_lib/src/DbClientImpl.cc
            orm_lib/src/DbClientLockFree.cc
            orm_lib/src/DbConnection.cc
            orm_lib/src/Exception.cc
            orm_lib/src/Field.cc
            orm_lib/src/Result.cc
            orm_lib/src/Row.cc
            orm_lib/src/SqlBinder.cc
            orm_lib/src/TransactionImpl.cc
            orm_lib/src/RestfulController.cc)
if (pg_FOUND OR MySQL_FOUND OR SQLite3_FOUND)
    set(DROGON_SOURCES
            ${DROGON_SOURCES}
            orm_lib/src/DbClientManager.cc)
else (pg_FOUND OR MySQL_FOUND OR SQLite3_FOUND)
    set(DROGON_SOURCES ${DROGON_SOURCES} lib/src/DbClientManagerSkipped.cc)
endif (pg_FOUND OR MySQL_FOUND OR SQLite3_FOUND)
```

Reply:
It works for orm ! :tada: Thanks

Reply:
I updated in c71180b I still need to do equivalent for Redis (same does not work as .cc files directly import `hiredis/hiredis.h`)